v03_pipeline/migrations/lookup/0003_add_sample_type.py

```diff
@@ -0,0 +1,93 @@
+import hail as hl
+import hailtop.fs as hfs
+
+from v03_pipeline.lib.logger import get_logger
+from v03_pipeline.lib.migration.base_migration import BaseMigration
+from v03_pipeline.lib.model import DatasetType, ReferenceGenome, SampleType
+from v03_pipeline.lib.paths import project_table_path
+
+logger = get_logger(__name__)
+
+
+class AddLookupSampleType(BaseMigration):
+    reference_genome_dataset_types: frozenset[
+        tuple[ReferenceGenome, DatasetType]
+    ] = frozenset(
+        (
+            (ReferenceGenome.GRCh37, DatasetType.SNV_INDEL),
+            (ReferenceGenome.GRCh38, DatasetType.SNV_INDEL),
+            (ReferenceGenome.GRCh38, DatasetType.MITO),
+        ),
+    )
+
+    @staticmethod
+    def migrate(
+        ht: hl.Table,
+        reference_genome: ReferenceGenome,
+        dataset_type: DatasetType,
+    ) -> hl.Table:
+        """
+        Adds sample_type to lookup ht global fields project_guids and project_families.
+        Assumes that only one project_ht exists for each project across both sample types.
+
+        Old Global fields:
+            'project_guids': array<str>
+            'project_families': dict<str, array<str>>
+            'updates': set<struct {
+                callset: str,
+                project_guid: str,
+                remap_pedigree_hash: int32
+            }>
+        New Global fields:
+            'project_guids': array<tuple (
+                str,
+                str
+            )>
+            'project_families': dict<tuple (
+                str,
+                str
+            ), array<str>>
+            'updates': set<struct {
+                callset: str,
+                project_guid: str,
+                remap_pedigree_hash: int32
+            }>
+        """
+        collected_globals = ht.globals.collect()[0]
+        project_guids = collected_globals['project_guids']
+        project_families = collected_globals['project_families']
+        projects_without_hts = set()
+
+        for i, project_guid in enumerate(project_guids):
+            project_has_ht = False
+            for sample_type in SampleType:
+                project_ht_path = project_table_path(
+                    reference_genome,
+                    dataset_type,
+                    sample_type,
+                    project_guid,
+                )
+                if not hfs.exists(project_ht_path):
+                    continue
+
+                project_has_ht = True
+                key = (project_guid, sample_type.value)
+                project_guids[i] = key
+                project_families[key] = project_families.pop(project_guid)
+                break
+
+            # It is possible that there are projects in the lookup globals with no corresponding project ht.
+            # For those projects, set sample_type to missing and log project_guid, so we can delete them later.
```

Comment:
This feels like a case that should be impossible. Would it be better for this to hard fail, and for us to first have to run any project deletions before we can run the migration?

Reply:
I agree with this, hard fail makes sense and if there's something out of wack we can debug outside of this migration.

Reply:
so I ran the migration without actually applying it with this check, here are projects that appear in project_guids but have no project tables:
MITO: R0566_tgg_ravenscroft_wes
SNV_INDEL 38: 52224_julia_test
SNV_INDEL 37: 52124_julia_test

They need to be removed from the lookup table.
- However, I don't want to run the delete project dag for 52224_julia_test since it has a family in project_families (['F051011_fam2']) and I don't want that family table to be deleted as well! 
- R0566_tgg_ravenscroft_wes I'm not sure about. It also has families in project_families but the corresponding family tables don't exist.

Reply:
ah yes.  `R0566_tgg_ravenscroft_wes` is my fault.

Reply:
We should run the `delete` families task... The family and project table didn't get copied over during the migration.

Reply:
R0566_tgg_ravenscroft_wes is gone from MITO
52124_julia_test is gone from SNV 37
still waiting on 52224_julia_test to be deleted, got a notebook running now in a cluster